lib/utility/sink.rb

```diff
@@ -12,28 +12,57 @@
 require 'utility/es_client'
 
 module Utility
-  module Sink
-    module_function
+  class Batcher
+    def initialize(sink, flush_threshold = 50)
+      # threshold and size are the same now, but it's good to separate the concepts:
+      # threshold is "when to start flushing", while "size" is the size of the batch
+      # that will be sent to the sink
+      @flush_threshold = flush_threshold
+      @flush_size = flush_threshold
+      @sink = sink
+      @queue = []
+    end
+
+    def add(element)
+      @queue << element
+
+      flush if ready_to_flush?
+    end
+
+    def add_multiple(elements)
+      elements.each do |element|
+        @queue << element
+      end
 
-    def print_delim
-      puts '----------------------------------------------------'
+      flush if ready_to_flush?
     end
 
-    def print_header(header)
-      print_delim
-      puts header
-      print_delim
+    def flush
+      until @queue.empty?
+        items_to_ingest = @queue.pop(@flush_size)
+        @sink.ingest_multiple(items_to_ingest)
+      end
     end
 
-    class ConsoleSink
-      include Sink
+    private
 
-      def ingest(_document)
-        print_header 'Got a single document:'
+    def ready_to_flush?
+      @queue.size >= @flush_threshold
+    end
+  end
+
+  module Sink
+    class Base
+      def with_batching(batch_size = 50)
```

Comment:
Where is this used?

Reply:
My bad - it's not used in this PR actually, but I use it in my MongoDB PR that's in draft mode.

To use this abstraction in this PR we need to update the code of Gitlab connector, which should probably be done in a separate PR

Reply:
Which also brings in the question - do we want to give user the flexibility of choosing whether they want to batch the requests or not. I think it's nice to have, but it also increases the complexity of the system - more stuff to document and explain

Reply:
I'd say batching isn't a feature, it's a performance-related implementation detail. Since we always prefer to improve performance, we can safely make it a default implementation. For now, I can hardly imagine a case when an override would be needed, but if it happens we'll provide for it.

Reply:
Yeah I agree, though at current state of code it's probably not possible yet to have a robust transparent batching interface. I'd vote for merging this PR and taking additional time to make this batching interface transparent (probably not much time to do, just a lot of code moving around)